src/Features/JsonPatch.SystemTextJson/src/Internal/IAdapter.cs

```diff
@@ -0,0 +1,54 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Text.Json;
+
+namespace Microsoft.AspNetCore.JsonPatch.SystemTextJson.Internal;
+
+/// <summary>
+/// This API supports infrastructure and is not intended to be used
+/// directly from your code. This API may change or be removed in future releases.
+/// </summary>
+public interface IAdapter
+{
+    bool TryTraverse(
+        object target,
+        string segment,
+        JsonSerializerOptions serializerOptions,
+        out object nextTarget,
+        out string errorMessage);
+
+    bool TryAdd(
+        object target,
+        string segment,
+        JsonSerializerOptions serializerOptions,
+        object value,
+        out string errorMessage);
+
+    bool TryRemove(
+        object target,
+        string segment,
+        JsonSerializerOptions serializerOptions,
+        out string errorMessage);
+
+    bool TryGet(
+        object target,
+        string segment,
+        JsonSerializerOptions serializerOptions,
+        out object value,
+        out string errorMessage);
```

Comment:
Consider adding an `out Type memberType` here if you decide to go the route of removing `GetType` and using the S.T.J. metadata. (Or even `out JsonPropertyInfo propInfo`.)

Reply:
Discussed that this will cause a behavior change, so not going to make it. thanks again.